2.9BSD/usr/man/cat1/stty.1

STTY(1)             UNIX Programmer's Manual              STTY(1)

NAME
     stty - set terminal options

SYNOPSIS
     stty [ option ... ]

DESCRIPTION
     _S_t_t_y sets certain I/O options on the current output termi-
     nal, placing its output on the diagnostic output.  With no
     argument, it reports the speed of the terminal and the set-
     tings of the options which are different from their
     defaults.  With the argument ``all'', all normally used
     option settings are reported.  With the argument ``every-
     thing'', everything _s_t_t_y knows about is printed.  The option
     strings are selected from the following set:

     even      allow even parity input
     -even     disallow even parity input
     odd       allow odd parity input
     -odd      disallow odd parity input
     raw       raw mode input (no input processing (erase, kill,
               interrupt, ...); parity bit passed back)
     -raw      negate raw mode
     cooked    same as `-raw'
     cbreak    make each character available to _r_e_a_d(2) as
               received; no erase and kill processing, but all
               other processing (interrupt, suspend, ...) is per-
               formed
     -cbreak   make characters available to _r_e_a_d only when new-
               line is received
     -nl       allow carriage return for new-line, and output
               CR-LF for carriage return or new-line
     nl        accept only new-line to end lines
     echo      echo back every character typed
     -echo     do not echo characters
     lcase     map upper case to lower case
     -lcase    do not map case
     tandem    enable flow control, so that the system sends out
               the stop character when its internal queue is in
               danger of overflowing on input, and sends the
               start character when it is ready to accept further
               input
     -tandem   disable flow control
     -tabs     replace tabs by spaces when printing
     tabs      preserve tabs
     ek        set erase and kill characters to # and @
     For the following commands which take a character argument
     _c, you may also specify _c as the ``u'' or ``undef'', to set
     the value to be undefined.  A value of ``^x'', a 2 character
     sequence, is also interpreted as a control character, with
     ``^?'' representing delete.

Printed 8/5/83                                                  1

STTY(1)             UNIX Programmer's Manual              STTY(1)

     erase _c   set erase character to _c (default `#', but often
               reset to ^H.)
     kill _c    set kill character to _c (default `@', but often
               reset to ^U.)
     intr _c    set interrupt character to _c (default DEL or ^?
               (delete), but often reset to ^C.)
     quit _c    set quit character to _c (default control \.)
     start _c   set start character to _c (default control Q.)
     stop _c    set stop character to _c (default control S.)
     eof _c     set end of file character to _c (default control
               D.)
     brk _c     set break character to _c (default undefined.) This
               character is an extra wakeup causing character.
     cr0 cr1 cr2 cr3
               select style of delay for carriage return (see
               _i_o_c_t_l(2))
     nl0 nl1 nl2 nl3
               select style of delay for linefeed
     tab0 tab1 tab2 tab3
               select style of delay for tab
     ff0 ff1   select style of delay for form feed
     bs0 bs1   select style of delay for backspace

     tty33     set all modes suitable for the Teletype Corpora-
               tion Model 33 terminal.
     tty37     set all modes suitable for the Teletype Corpora-
               tion Model 37 terminal.
     vt05      set all modes suitable for Digital Equipment Corp.
               VT05 terminal
     dec       set all modes suitable for Digital Equipment Corp.
               operating systems users; (erase, kill, and inter-
               rupt characters to ^?, ^U, and ^C, decctlq and
               ``newcrt''.)

     tn300     set all modes suitable for a General Electric Ter-
               miNet 300
     ti700     set all modes suitable for Texas Instruments 700
               series terminal
     tek       set all modes suitable for Tektronix 4014 terminal
     0         hang up phone line immediately
     50 75 110 134 150 200 300 600 1200 1800 2400 4800 9600 exta extb
               Set terminal baud rate to the number given, if
               possible.  (These are the speeds supported by the
               DH-11 interface).

     A teletype driver which supports the job control processing
     of _c_s_h(1) as introduced in _n_e_w_c_s_h(1) and more functionality
     that the basic driver is introduced in _n_e_w_t_t_y(4) and fully
     described in _t_t_y(4).  The following options apply only to
     it.

     new       Use new driver (switching flushes typeahead).

Printed 8/5/83                                                  2

STTY(1)             UNIX Programmer's Manual              STTY(1)

     crt       Set options for a CRT (crtbs, ctlecho and, if >=
               1200 baud, crterase and crtkill.)
     crtbs     Echo backspaces on erase characters.
     prterase  For printing terminal echo erased characters back-
               wards within ``\'' and ``/''.
     crterase  Wipe out erased characters with ``backspace-
               space-backspace.''
     -crterase Leave erased characters visible; just backspace.
     crtkill   Wipe out input on like kill ala crterase.
     -crtkill  Just echo line kill character and a newline on
               line kill.
     ctlecho   Echo control characters as ``^_x'' (and delete as
               ``^?''.) Print two backspaces following the EOT
               character (control D).
     -ctlecho  Control characters echo as themselves; in cooked
               mode EOT (control-D) is not echoed.

     decctlq   After output is suspended (normally by ^S), only a
               start character (normally ^Q) will restart it.
               This is compatible with DEC's vendor supplied sys-
               tems.

     -decctlq  After output is suspended, any character typed
               will restart it; the start character will restart
               output without providing any input.  (This is the
               default.)
     tostop    Background jobs stop if they attempt terminal out-
               put.
     -tostop   Output from background jobs to the terminal is
               allowed.
     tilde     Convert ``~'' to ```'' on output (for Hazeltine
               terminals).
     -tilde    Leave poor ``~'' alone.
     flusho    Output is being discarded usually because user hit
               control O (internal state bit).
     -flusho   Output is not being discarded.
     pendin    Input is pending after a switch from cbreak to
               cooked and will be re-input when a read becomes
               pending or more input arrives (internal state
               bit).
     -pendin   Input is not pending.
     intrup    Send a signal (SIGTINT) to the terminal control
               process group whenever an input record (line in
               cooked mode, character in cbreak or raw mode) is
               available for reading.
     -intrup   Don't send input available interrupts.
     mdmbuf    Start/stop output on carrier transitions (not
               implemented).
     -mdmbuf   Return error if write attempted after carrier
               drops.
     litout    Send output characters without any processing.
     -litout   Do normal output processing, inserting delays,

Printed 8/5/83                                                  3

STTY(1)             UNIX Programmer's Manual              STTY(1)

               etc.
     nohang    Don't send hangup signal if carrier drops.
     -nohang   Send hangup signal to control process group when
               carrier drops.
     etxack    Diablo style etx/ack handshaking (not imple-
               mented).

     The following special characters are applicable only to the
     new teletype driver and are not normally changed.

     susp _c    set suspend process character to _c (default con-
               trol Z.)
     dsusp _c   set delayed suspend process character to _c
               (default control Y.)
     rprnt _c   set reprint line character to _c (default control
               R.)
     flush _c   set flush output character to _c (default control
               O.)
     werase _c  set word erase character to _c (default control W.)
     lnext _c   set literal next character to _c (default control
               V.)

SEE ALSO
     tabs(1), tset(1), ioctl(2), newtty(4), tty(4)

Printed 8/5/83                                                  4



2.9BSD/usr/src/cmd/stty.c
#ifndef	lint
static char *sccsid ="@(#)stty.c	4.7 (Berkeley) 7/18/81";
#endif
/*
 * set teletype modes
 */

#include <whoami.h>
#include <stdio.h>
#include <sgtty.h>

struct
{
	char	*string;
	int	speed;
} speeds[] = {
	"0",	B0,
	"50",	B50,
	"75",	B75,
	"110",	B110,
	"134",	B134,
	"134.5",B134,
	"150",	B150,
	"200",	B200,
	"300",	B300,
	"600",	B600,
	"1200",	B1200,
	"1800",	B1800,
	"2400",	B2400,
	"4800",	B4800,
	"9600",	B9600,
	"exta",	EXTA,
	"extb",	EXTB,
	0,
};
struct
{
	char	*string;
	int	set;
	int	reset;
	int	lset;
	int	lreset;
} modes[] = {
	"even",		EVENP, 0, 0, 0,
	"-even",	0, EVENP, 0, 0,
	"odd",		ODDP, 0, 0, 0,
	"-odd",		0, ODDP, 0, 0,
	"raw",		RAW, 0, 0, 0,
	"-raw",		0, RAW, 0, 0,
	"cooked",	0, RAW, 0, 0,
	"-nl",		CRMOD, 0, 0, 0,
	"nl",		0, CRMOD, 0, 0,
	"echo",		ECHO, 0, 0, 0,
	"-echo",	0, ECHO, 0, 0,
	"LCASE",	LCASE, 0, 0, 0,
	"lcase",	LCASE, 0, 0, 0,
	"-LCASE",	0, LCASE, 0, 0,
	"-lcase",	0, LCASE, 0, 0,
	"-tabs",	XTABS, 0, 0, 0,
	"tabs",		0, XTABS, 0, 0,
	"tandem",	TANDEM, 0, 0, 0,
	"-tandem",	0, TANDEM, 0, 0,
	"cbreak",	CBREAK, 0, 0, 0,
	"-cbreak",	0, CBREAK, 0, 0,
	"cr0",		CR0, CR3, 0, 0,
	"cr1",		CR1, CR3, 0, 0,
	"cr2",		CR2, CR3, 0, 0,
	"cr3",		CR3, CR3, 0, 0,
	"tab0",		TAB0, XTABS, 0, 0,
	"tab1",		TAB1, XTABS, 0, 0,
	"tab2",		TAB2, XTABS, 0, 0,
	"nl0",		NL0, NL3, 0, 0,
	"nl1",		NL1, NL3, 0, 0,
	"nl2",		NL2, NL3, 0, 0,
	"nl3",		NL3, NL3, 0, 0,
	"ff0",		FF0, FF1, 0, 0,
	"ff1",		FF1, FF1, 0, 0,
	"bs0",		BS0, BS1, 0, 0,
	"bs1",		BS1, BS1, 0, 0,
	"33",		CR1, ALLDELAY, 0, 0,
	"tty33",	CR1, ALLDELAY, 0, 0,
	"37",		FF1+CR2+TAB1+NL1, ALLDELAY, 0, 0,
	"tty37",	FF1+CR2+TAB1+NL1, ALLDELAY, 0, 0,
	"05",		NL2, ALLDELAY, 0, 0,
	"vt05",		NL2, ALLDELAY, 0, 0,
	"tn",		CR1, ALLDELAY, 0, 0,
	"tn300",	CR1, ALLDELAY, 0, 0,
	"ti",		CR2, ALLDELAY, 0, 0,
	"ti700",	CR2, ALLDELAY, 0, 0,
	"tek",		FF1, ALLDELAY, 0, 0,
	"crtbs",	0, 0, LCRTBS, LPRTERA,
	"-crtbs",	0, 0, 0, LCRTBS,
	"prterase",	0, 0, LPRTERA, LCRTBS+LCRTKIL+LCRTERA,
	"-prterase",	0, 0, 0, LPRTERA,
	"crterase",	0, 0, LCRTERA, LPRTERA,
	"-crterase",	0, 0, 0, LCRTERA,
	"crtkill",	0, 0, LCRTKIL, LPRTERA,
	"-crtkill",	0, 0, 0, LCRTKIL,
	"tilde",	0, 0, LTILDE, 0,
	"-tilde",	0, 0, 0, LTILDE,
	"mdmbuf",	0, 0, LMDMBUF, 0,
	"-mdmbuf",	0, 0, 0, LMDMBUF,
	"litout",	0, 0, LLITOUT, 0,
	"-litout",	0, 0, 0, LLITOUT,
	"tostop",	0, 0, LTOSTOP, 0,
	"-tostop",	0, 0, 0, LTOSTOP,
	"flusho",	0, 0, LFLUSHO, 0,
	"-flusho",	0, 0, 0, LFLUSHO,
	"nohang",	0, 0, LNOHANG, 0,
	"-nohang",	0, 0, 0, LNOHANG,
	"etxack",	0, 0, LETXACK, 0,
	"-etxack",	0, 0, 0, LETXACK,
	"intrup",	0, 0, LINTRUP, 0,
	"-intrup",	0, 0, 0, LINTRUP,
	"ctlecho",	0, 0, LCTLECH, 0,
	"-ctlecho",	0, 0, 0, LCTLECH,
	"pendin",	0, 0, LPENDIN, 0,
	"-pendin",	0, 0, 0, LPENDIN,
	"decctlq",	0, 0, LDECCTQ, 0,
	"-decctlq",	0, 0, 0, LDECCTQ,
	"noflsh",	0, 0, LNOFLSH, 0,
	"-noflsh",	0, 0, 0, LNOFLSH,
	0,
};

struct tchars tc;
struct ltchars ltc;
struct sgttyb mode;
int	lmode;
int	oldisc, ldisc;

#define	CTRL(x)		('x'&037)

struct	special {
	char	*name;
	char	*cp;
	char	def;
} special[] = {
	"erase",	&mode.sg_erase,		CTRL(h),
	"kill",		&mode.sg_kill,		'@',
	"intr",		&tc.t_intrc,		0177,
	"quit",		&tc.t_quitc,		CTRL(\\\\),
	"start",	&tc.t_startc,		CTRL(q),
	"stop",		&tc.t_stopc,		CTRL(s),
	"eof",		&tc.t_eofc,		CTRL(d),
	"brk",		&tc.t_brkc,		0377,
	"susp",		&ltc.t_suspc,		CTRL(z),
	"dsusp",	&ltc.t_dsuspc,		CTRL(y),
	"rprnt",	&ltc.t_rprntc,		CTRL(r),
	"flush",	&ltc.t_flushc,		CTRL(o),
	"werase",	&ltc.t_werasc,		CTRL(w),
	"lnext",	&ltc.t_lnextc,		CTRL(v),
	0
};
char	*arg;

int	argc;
char	**argv;
main(iargc, iargv)
char	**iargv;
{
	int i;
	register struct special *sp;
	char obuf[BUFSIZ];

	setbuf(stderr, obuf);
	argc = iargc;
	argv = iargv;
	gtty(1, &mode);
	ioctl(1, TIOCGETD, (struct sgttyb *) &ldisc);
	oldisc = ldisc;
	ioctl(1, TIOCGETC, (struct sgttyb *) &tc);
	ioctl(1, TIOCLGET, (struct sgttyb *) &lmode);
	ioctl(1, TIOCGLTC, (struct sgttyb *) &ltc);
	if(argc == 1) {
		prmodes(0);
		exit(0);
	}
	if (argc == 2 && !strcmp(argv[1], "all")) {
		prmodes(1);
		exit(0);
	}
	if (argc == 2 && !strcmp(argv[1], "everything")) {
		prmodes(2);
		exit(0);
	}
/*
	if (argc == 2 && !strcmp(argv[1], "all")) {
		prmodes(2);
		exit(0);
	}
*/
	while(--argc > 0) {
		arg = *++argv;
		if (eq("ek")){
			mode.sg_erase = '#';
			mode.sg_kill = '@';
			continue;
		}
		if (eq("new")){
			ldisc = NTTYDISC;
			if (ioctl(1, TIOCSETD, (struct sgttyb *) &ldisc)<0)
				perror("ioctl");
			continue;
		}
		if (eq("newcrt")){
			ldisc = NTTYDISC;
			lmode &= ~LPRTERA;
			lmode |= LCRTBS|LCTLECH;
			if (mode.sg_ospeed >= B1200)
				lmode |= LCRTERA|LCRTKIL;
			if (ioctl(1, TIOCSETD, (struct sgttyb *) &ldisc)<0)
				perror("ioctl");
			continue;
		}
		if (eq("crt")){
			lmode &= ~LPRTERA;
			lmode |= LCRTBS|LCTLECH;
			if (mode.sg_ospeed >= B1200)
				lmode |= LCRTERA|LCRTKIL;
			continue;
		}
		if (eq("old")){
			ldisc = OTTYDISC;
			if (ioctl(1, TIOCSETD, (struct sgttyb *) &ldisc)<0)
				perror("ioctl");
			continue;
		}
		if (eq("dec")){
			mode.sg_erase = 0177;
			mode.sg_kill = CTRL(u);
			tc.t_intrc = CTRL(c);
			ldisc = NTTYDISC;
			lmode &= ~LPRTERA;
			lmode |= LCRTBS|LCTLECH|LDECCTQ;
			if (mode.sg_ospeed >= B1200)
				lmode |= LCRTERA|LCRTKIL;
			if (ioctl(1, TIOCSETD, (struct sgttyb *) &ldisc)<0)
				perror("ioctl");
			continue;
		}
		for (sp = special; sp->name; sp++)
			if (eq(sp->name)) {
				if (--argc == 0)
					goto done;
				if (**++argv == 'u')
					*sp->cp = 0377;
				else if (**argv == '^')
					*sp->cp = ((*argv)[1] == '?') ?
					    0177 : (*argv)[1] & 037;
				else
					*sp->cp = **argv;
				goto cont;
			}
		if (eq("gspeed")) {
			mode.sg_ispeed = B300;
			mode.sg_ospeed = B9600;
			continue;
		}
		if (eq("hup")) {
			ioctl(1, TIOCHPCL, (struct sgttyb *) NULL);
			continue;
		}
		for(i=0; speeds[i].string; i++)
			if(eq(speeds[i].string)) {
				mode.sg_ispeed = mode.sg_ospeed = speeds[i].speed;
				goto cont;
			}
		if (eq("speed")) {
			gtty(open("/dev/tty", 0), &mode);
			for(i=0; speeds[i].string; i++)
				if (mode.sg_ospeed == speeds[i].speed) {
					printf("%s\n", speeds[i].string);
					exit(0);
				}
			printf("unknown\n");
			exit(1);
		}
		for(i=0; modes[i].string; i++)
			if(eq(modes[i].string)) {
				mode.sg_flags &= ~modes[i].reset;
				mode.sg_flags |= modes[i].set;
				lmode &= ~modes[i].lreset;
				lmode |= modes[i].lset;
			}
		if(arg)
			fprintf(stderr,"unknown mode: %s\n", arg);
cont:
		;
	}
done:
	ioctl(1, TIOCSETN, (struct sgttyb *) &mode);
	ioctl(1, TIOCSETC, (struct sgttyb *) &tc);
	ioctl(1, TIOCSLTC, (struct sgttyb *) &ltc);
	ioctl(1, TIOCLSET, (struct sgttyb *) &lmode);
}

eq(string)
char *string;
{
	int i;

	if(!arg)
		return(0);
	i = 0;
loop:
	if(arg[i] != string[i])
		return(0);
	if(arg[i++] != '\0')
		goto loop;
	arg = 0;
	return(1);
}

prmodes(all)
{
	register m;
	int any;

	if(ldisc==NETLDISC)
		fprintf(stderr, "net discipline, ");
	else if(ldisc==NTTYDISC)
		fprintf(stderr, "new tty, ");
	else if(all==2)
		fprintf(stderr, "old tty, ");
	if(mode.sg_ispeed != mode.sg_ospeed) {
		prspeed("input speed ", mode.sg_ispeed);
		prspeed("output speed ", mode.sg_ospeed);
	} else
		prspeed("speed ", mode.sg_ispeed);
	fprintf(stderr, all==2 ? "\n" : "; ");
	m = mode.sg_flags;
	if(all==2 || (m&(EVENP|ODDP))!=(EVENP|ODDP)) {
		if(m & EVENP)	fprintf(stderr,"even ");
		if(m & ODDP)	fprintf(stderr,"odd ");
	}
	if(all==2 || m&RAW)
		fprintf(stderr,"-raw "+((m&RAW)!=0));
	if(all==2 || (m&CRMOD)==0)
		fprintf(stderr,"-nl "+((m&CRMOD)==0));
	if(all==2 || (m&ECHO)==0)
		fprintf(stderr,"-echo "+((m&ECHO)!=0));
	if(all==2 || (m&LCASE))
		fprintf(stderr,"-lcase "+((m&LCASE)!=0));
	if(all==2 || (m&TANDEM))
		fprintf(stderr,"-tandem "+((m&TANDEM)!=0));
	fprintf(stderr,"-tabs "+((m&XTABS)!=XTABS));
	if(all==2 || (m&CBREAK))
		fprintf(stderr,"-cbreak "+((m&CBREAK)!=0));
	if(all==2 || (m&NLDELAY))
		delay((m&NLDELAY)/NL1,	"nl");
	if ((m&TBDELAY)!=XTABS)
		delay((m&TBDELAY)/TAB1,	"tab");
	if(all==2 || (m&CRDELAY))
		delay((m&CRDELAY)/CR1,	"cr");
	if(all==2 || (m&VTDELAY))
		delay((m&VTDELAY)/FF1,	"ff");
	if(all==2 || (m&BSDELAY))
		delay((m&BSDELAY)/BS1,	"bs");
	if (all)
		fprintf(stderr,"\n");
#define	lpit(what,str) \
	if (all==2||(lmode&what)) { \
		fprintf(stderr,str+((lmode&what)!=0)); any++; \
	}
	if (ldisc == NTTYDISC) {
		int newcrt = (lmode&(LCTLECH|LCRTBS)) == (LCTLECH|LCRTBS) &&
		    (lmode&(LCRTERA|LCRTKIL)) ==
		      ((mode.sg_ospeed > B300) ? LCRTERA|LCRTKIL : 0);
		if (newcrt) {
			if (all==2)
				fprintf(stderr, "crt: (crtbs crterase crtkill ctlecho) ");
			else
				fprintf(stderr, "crt ");
			any++;
		} else {
			lpit(LCRTBS, "-crtbs ");
			lpit(LCRTERA, "-crterase ");
			lpit(LCRTKIL, "-crtkill ");
			lpit(LCTLECH, "-ctlecho ");
			lpit(LPRTERA, "-prterase ");
		}
		lpit(LTOSTOP, "-tostop ");
		lpit(LINTRUP, "-intrup ");
		if (all==2) {
			fprintf(stderr, "\n");
			any = 0;
		}
		lpit(LTILDE, "-tilde ");
		lpit(LFLUSHO, "-flusho ");
		lpit(LMDMBUF, "-mdmbuf ");
		lpit(LLITOUT, "-litout ");
		lpit(LNOHANG, "-nohang ");
		if (any) {
			fprintf(stderr,"\n");
			any = 0;
		}
		lpit(LETXACK, "-etxack ");
		lpit(LPENDIN, "-pendin ");
		lpit(LDECCTQ, "-decctlq ");
		lpit(LNOFLSH, "-noflsh ");
		if (any)
			fprintf(stderr,"\n");
	} else if (!all)
		fprintf(stderr,"\n");
	if (all) {
		switch (ldisc) {

		case OTTYDISC:
			fprintf(stderr,"\
erase  kill   intr   quit   stop   eof\
\n");
			pcol(mode.sg_erase, -1);
			pcol(mode.sg_kill, -1);
			pcol(tc.t_intrc, -1);
			pcol(tc.t_quitc, -1);
			pcol(tc.t_stopc, tc.t_startc);
			pcol(tc.t_eofc, tc.t_brkc);
			fprintf(stderr,"\n");
			break;

		case NTTYDISC:
			fprintf(stderr,"\
erase  kill   werase rprnt  flush  lnext  susp   intr   quit   stop   eof\
\n"); 
			pcol(mode.sg_erase, -1);
			pcol(mode.sg_kill, -1);
			pcol(ltc.t_werasc, -1);
			pcol(ltc.t_rprntc, -1);
			pcol(ltc.t_flushc, -1);
			pcol(ltc.t_lnextc, -1);
			pcol(ltc.t_suspc, ltc.t_dsuspc);
			pcol(tc.t_intrc, -1);
			pcol(tc.t_quitc, -1);
			pcol(tc.t_stopc, tc.t_startc);
			pcol(tc.t_eofc, tc.t_brkc);
			fprintf(stderr,"\n");
			break;
		}
	} else if (ldisc != NETLDISC) {
		register struct special *sp;
		int first = 1;
		for (sp = special; sp->name; sp++) {
			if ((*sp->cp&0377) != (sp->def&0377)) {
				pit((unsigned) (*sp->cp), sp->name, first ? "" : ", ");
				first = 0;
			};
			if (sp->cp == &tc.t_brkc && ldisc == OTTYDISC)
				break;
		}
		if (first == 0)
			fprintf(stderr, "\n");
	}
}

pcol(ch1, ch2)
	int ch1, ch2;
{
	int nout = 0;

	ch1 &= 0377;
	ch2 &= 0377;
	if (ch1 == ch2)
		ch2 = 0377;
	for (; ch1 != 0377 || ch2 != 0377; ch1 = ch2, ch2 = 0377) {
		if (ch1 == 0377)
			continue;
		if (ch1 & 0200) {
			fprintf(stderr, "M-");
			nout += 2;
			ch1 &= ~ 0200;
		}
		if (ch1 == 0177) {
			fprintf(stderr, "^");
			nout++;
			ch1 = '?';
		} else if (ch1 < ' ') {
			fprintf(stderr, "^");
			nout++;
			ch1 += '@';
		}
		fprintf(stderr, "%c", ch1);
		nout++;
		if (ch2 != 0377) {
			fprintf(stderr, "/");
			nout++;
		}
	}
	while (nout < 7) {
		fprintf(stderr, " ");
		nout++;
	}
}

pit(what, itsname, sep)
	unsigned what;
	char *itsname, *sep;
{

	what &= 0377;
	fprintf(stderr, "%s%s", sep, itsname);
	if (what == 0377) {
		fprintf(stderr, " <undef>");
		return;
	}
	fprintf(stderr, " = ");
	if (what & 0200) {
		fprintf(stderr, "M-");
		what &= ~ 0200;
	}
	if (what == 0177) {
		fprintf(stderr, "^");
		what = '?';
	} else if (what < ' ') {
		fprintf(stderr, "^");
		what += '@';
	}
	fprintf(stderr, "%c", what);
}

delay(m, s)
char *s;
{

	if(m)
		fprintf(stderr,"%s%d ", s, m);
}

int	speed[] = {
	0,50,75,110,134,150,200,300,600,1200,1800,2400,4800,9600,0,0
};

prspeed(c, s)
char *c;
{

	fprintf(stderr,"%s%d baud",  c, speed[s]);
}
